src/System.Linq/src/System/Linq/Zip.cs

```diff
@@ -28,6 +28,9 @@ public static partial class Enumerable
             return ZipIterator(first, second, resultSelector);
         }
 
+        public static IEnumerable<(TFirst, TSecond)> Zip<TFirst, TSecond>(this IEnumerable<TFirst> first, IEnumerable<TSecond> second)
```

Comment:
Do we want to name the TFirst and TSecond in the result tuple?  Otherwise they only show up in IntelliSense as "Item1" and "Item2".  Should we name them "First" and "Second"?

Reply:
@terrajobst, @KrzysztofCwalina, @bartonjs, have we thought through design guidelines for public APIs exposing tuples?  IIRC, this is the first time we're doing it.

Reply:
Unless you really feel `first` is more meaningful than `Item1` I suggest you don't include element names in the public API. In the case of Zip I'd just let the user name the tuple elements, e.g.:

```cs
foreach (var (name, age) in names.Zip(ages))
{
    ...
}
```

Reply:
I believe our design guideline for Tuple/ValueTuple is "AVOID".

Reply:
> I believe our design guideline for Tuple/ValueTuple is "AVOID".

Apparently we approved this one, though?

Reply:
I tracked Krzysztof down and he expressed the opinion `(TFirst First, TSecond Second)` (property-cased, since that's how the caller sees them).  I agree with it, and post-facto feel that the ValueTuple return from Zip sort of reasonable given that it's inherently just a tuple factory by design (and a generic-composed one at that).

Reply:
Roslyn has been tending more towards local case, because that's often how they're being deconstructed. Is this the first instance of tuple names in CoreFX?

Reply:
> Is this the first instance of tuple names in CoreFX?

In public API, yes, I believe so.

Reply:
It looks like the IDE defaults to camelCase. I think we should go with that unless there's significant reason to deviate from VS guidelines.

Reply:
If we were just using the existing properties, it would be Item1 and Item2, so PascalCased.

If we were defining a custom return struct here, it would be First and Second properties, so PascalCased.

It seems strange to me then that we'd deviate from that just because we're using a language feature to avoid creating a custom struct type.

I personally don't think the IDE's guidance is correct for public APIs. For internal usage, anything goes.

Reply:
As far as coding guidelines go, I don't think "correct" means anything -- the only thing that matters is consistency. If we want to publish a C# language style that recommends PascalCase and change VS to use that default, I'm fine with that.

Reply:
We re-reviewed this in our API review meeting this week, and we reiterated the property/pascal casing.

So the .NET API Guidance for this is, effectively (using Framework Design Guidelines syntax):

* AVOID using Tuple or ValueTuple in public API
* DO name ValueTuple members as if they were properties when using it in public API

("**Avoid** guidelines indicate something is generally not a good idea, but there are known cases where breaking the rule makes sense.")

Making VS suggest that by default seems reasonable, but that's not something I know how to direct.

Reply:
Cool, I'll forward this to the IDE team.

Reply:
😢 I thought the guidelines was always lowercase...

Reply:
Every single time tuples were shown as a new language feature by @MadsTorgersen it was lowercase. The idea was always to make them look like parameter lists, especially for when you're returning tuple.

> If we were defining a custom return struct here, it would be First and Second properties, so PascalCased.

@stephentoub Yes, but you're not defining a custom struct. The idea with tuples was that you shouldn't really think of them as their own thing or as representing any concept, it's just a way to group two variables together. Yes, the elements are really fields, but you shouldn't have to think of them as fields, you should think of them as just variables (after all that's why they ended up being mutable fields as opposed to properties).

This decision goes against everything I've heard on this subject until now. Why is the guideline changing after we've had tuples for more than a year...? :confused:

Reply:
> If we were defining a custom return struct here, it would be First and Second properties, so PascalCased.

If we were using out parameters, they'd be lowercase.

Reply:
> If we were using out parameters, they'd be lowercase.

But we're not.

> Yes, but you're not defining a custom struct.

We essentially are.

Reply:
> Why is the guideline changing after we've had tuples for more than a year...?

We've never before created any guidelines on this subject about the creation of public APIs in .NET libraries; can you please point me to where you've seen otherwise?  This is the first API in all of the .NET libraries where we've added a tuple return type, and as such we have now created that guideline, and it's that in public APIs, these should be PascalCased.

Reply:
We're not. I don't know what you mean by "essentially are", other than an analogy. in which case you could make an analogy with out variables as well.

Reply:
This API is returning in a single instance multiple pieces of data.  In *every* .NET API where that happens, those multiple pieces of data are exposed via PascalCased members off of that returned instance.  Here, it's happening via a returned struct, and rather than having to define that struct manually, we're using a language syntax that let's us do it inline.  That's what I meant by "essentially are": we're customizing the struct that's returned so that rather than having the PascalCased Item1 and Item2, we have the PascalCased First and Second.

Reply:
A tuple type is technically not a struct.

Reply:
You said this API is returning a struct. From a C# perspective, I don't think that is the case.

Reply:
I respectfully disagree.  I also don't see how that matters for naming.